### DISCUSSION TITLE
How to check for correctess while using units?

### DISCUSSION THREAD

**Discussion by jdacoello** [qudt/qudt-public-repo#473] [Q&A] [answered]
This question refers to the practical use of a `QuantityKind` and the expected `Unit`. According to the ontology, each individual of `QuantityKind` has asserted a list of multiple units that are expected via the `applicableUnit` object property.

For example, the quantity kind `velocity` has the following applicable units:
* `KiloM-PER_HR`
* `MI-PER_HR`
* etc...

I am wondering if, while using the ontology in practice, there is a way to avoid using the wrong unit.

For example:
A custom ontology has the concept of `Speed`. To add context to the values that are measured from this signal, the custom ontology reuses the qudt quantity kind of `velocity`

```turtle
eg:Speed qudt:hasQuantityKind qudt:Velocity .
```

What happens if someone by mistake uses the unit `DEG_C` (Degrees Celsius), which is of the kind `Temperature` instead of one of the expected units for `velocity`?
In other words, what is the proper way to check for correctness?

Thanks!

**Comment by steveraysteveray** [comment, marked as answer]
One quick way is to check the DimensionVector that a Unit points to, to check if it is the same as the DimensionVector pointed to by the QuantityKind. That at least tells you if the Unit and the QuantityKind are compatible.

Your example would fail that test.

**Reply by jdacoello** [reply]
I see. Then, @steveraysteveray is this test that you mention implemented..

- ..via SHACL?
- ..with a simple SPARQL query in the application that uses the data?
- ..other way?

Would you mind writing back a toy example?

Thank you!

**Reply by steveraysteveray** [reply]
Here's a simple SPARQL query to test unit:DEG_C. It's not efficient, with all those OPTIONAL statements, but gives you the idea.
```
PREFIX qk: <http://qudt.org/vocab/quantitykind/>
PREFIX qkdv: <http://qudt.org/vocab/dimensionvector/>
PREFIX qudt: <http://qudt.org/schema/qudt/>
PREFIX unit: <http://qudt.org/vocab/unit/>

SELECT  *
WHERE {
  BIND (unit:DEG_C as ?unit) .
  OPTIONAL {?unit qudt:hasQuantityKind ?qk .}
  OPTIONAL {?unit qudt:hasDimensionVector ?uVec .}
  OPTIONAL {?qk qudt:hasDimensionVector ?qVec .}
#  FILTER (?uVec != ?qVec) .
}
```
I commented out the FILTER statement so you could just see the values. You could turn this into a SHACL validation query by including the FILTER statement and putting the whole thing inside a SHACL shape.

**Reply by steveraysteveray** [reply]
Actually, I took another look with this query (no OPTIONAL statements) and found 5 offending situations, so thanks for leading me down this path!
```
SELECT  *
WHERE {
  ?unit a qudt:Unit .
  ?unit qudt:hasQuantityKind ?qk .
  ?unit qudt:hasDimensionVector ?uVec .
  ?qk qudt:hasDimensionVector ?qVec .
  FILTER (?uVec != ?qVec) .
}
```
Results:
![image](https://user-images.githubusercontent.com/1130189/155770511-8befa9ce-a5b4-4962-9c81-17469f231bbf.png)

**Comment by jhodgesatmb** [comment]
﻿Since applicableUnit will always return a valid unit for a quantitykind the temperature example would fail since Deg_C would not be in the applicableUnit list. The particular example of Speed does need to be addressed but it is an anomaly (hopefully). In general, SHACL could be used in an application to test for validity, but the vocabularies are internally consistent.

Jack

**Comment by jdacoello** [comment]
It seems that it is actually a good way to check for consistency within the ontology.

Thank you for the examples. I’ll try to use something similar within a SHACL shape.
